[draft] Move monotonicity info to live directly on `UnaryFunc`

bkirwi opened the pull request:
Followup to #19298, handling unary funcs.

There are a lot of monotone funcs! I suspect there are some remaining funcs that are monotone that I haven't annotated, but this should include:
- all the funcs that were previously annotated;
- easy-to-classify funcs like casts;
- some other functions that I feel like might be surprising if they don't work, like rounding.

At this point, since `is_monotone` returns `bool`, we can get rid of `Monotonic::Maybe` and similar.

### Motivation

See #19298 for context.

### Tips for reviewer

For correctness, it's especially important that func never returns true when the function is not monotonic. Let me know if there's any func you're not totally sure of!

It's less crucial that all monotonic functions are annotated, but still nice. Feel free to ask me to look at more functions or suggest new annotations as well, if something feels especially important for temporal filters.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  <!-- Reference the design in the description. -->
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  <!-- Ask in #team-cloud on Slack if you need help preparing the cloud PR. -->
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - <!-- Add release notes here or explicitly state that there are no user-facing behavior changes. -->

def- commented on src/expr/src/interpret.rs:
```diff
@@ -976,8 +801,78 @@ impl<'a> Interpreter for ColumnSpecs<'a> {
+        (left && left_pushdownable) || (right && right_pushdownable)
+    }
+
+    fn variadic(&self, func: &VariadicFunc, exprs: Vec<Self::Summary>) -> Self::Summary {
```
This function seems unused in any test. Seems interesting enough to cover.

def- commented on src/expr/src/scalar/func/impls/date.rs:
```diff
@@ -103,6 +106,10 @@ impl<'a> EagerUnaryFunc<'a> for ExtractDate {
     }
+
+    fn is_monotone(&self) -> bool {
+        most_significant_unit(self.0)
```
Most of these is_monotone functions seem uncovered, this one seems to do something interesting.

bkirwi replied:
I've added the `date` type to the equivalence test in `interpret.rs`, which should give us a little coverage.

And yeah, not much test coverage for the annotations. The nicest test to write for this would be some property test that, for every function, generated some random inputs and verified that the function was behaving monotonically if it claimed to be monotonic... but right now this sort of test is very expensive to write. (And the easy tests to write -- which just assert that a particular `is_monotone()` call returns true or whatever -- don't seem especially useful.)

def- reviewed (commented):
I triggered nightly, which looked fine: https://buildkite.com/materialize/nightlies/builds/2496#_ (same failures as on main branch)

Coverage information marked inline: https://buildkite.com/materialize/coverage/builds/121

aalexandrov commented on src/expr/src/scalar/func/impls/jsonb.rs:
```diff
@@ -33,6 +34,9 @@ sqlfunc!(
 
 sqlfunc!(
     #[sqlname = "jsonb_to_smallint"]
+    #[preserves_uniqueness = false]
```
I think the defaults in `macro_rules! sqlfunc` are
```rust
#[preserves_uniqueness = false]
#[inverse = None]
````
so I wonder why you are explicitly adding them everywhere.

bkirwi replied:
The issue is that (AFAICT) the macro needs a separate pattern for every possible combination of annotations. Naively, this means that adding a new annotation doubles the number of necessary patterns.

I figured it was better to provide some redundant annotations than to add additional patterns to the macro, but I don't feel strongly about it; happy to go the other way if y'all prefer it.

aalexandrov replied:
The `sqlfunc!` calls itself recursively top-down. With the current definition:

https://github.com/MaterializeInc/materialize/blob/ee747b414cdb3f726de72a7a4fa8bc1feb6881ba/src/expr/src/scalar/func/macros.rs#L17-L139

* the first application will just add the default `#[sqlname = stringify!($fn_name)]`,
* the second will add `#[preserves_uniqueness = false]` and `#[inverse = None]`,
* and so forth ...

If you use `rust-analyzer` you should have an "expand macro recursively" action and check what happens if you don't annotate your function explicitly with the defaults. I'm pretty sure that with your current changes you should be able to validate if this works.

aalexandrov replied:
I see where the problem is, I'll try to see if there is a more elegant solution that covers `N` independent sets of defaults where any `M < N` of those can be present in some arbitrary order.

bkirwi replied:
Asked in our internal #rust-help channel... we'll see if anyone has ideas!

aalexandrov replied:
I think you might be able to pattern match all attributes with `$(#[$name = $val])` but I have not tried it out.

aalexandrov commented on src/expr/src/interpret.rs:
```diff
@@ -896,6 +801,75 @@ impl<'a> Interpreter for ColumnSpecs<'a> {
     }
 }
 
+/// An interpreter that returns whether or not a particular expression is "pushdownable".
```
This is very useful, you might want to refer to the `Trace` docs in the module-level docs of this module.

aalexandrov reviewed (approved):
Looks good, but I didn't do thorough checks for correctness of the annotations. I think some `proptest!` based approach should work best for checking wrongly specified annotations.

My only concern is duplicating the

```
#[preserves_uniqueness = false]
#[inverse = None]
```

calls across the board.

ggevay commented on src/expr/src/scalar/func/impls/timestamp.rs:
```diff
@@ -257,6 +278,10 @@ impl<'a> EagerUnaryFunc<'a> for ExtractTimestampTz {
+
+    fn is_monotone(&self) -> bool {
+        most_significant_unit(self.0)
+    }
```
(I'm not sure how this treats the offset, but might be suspect.)

bkirwi commented on src/expr/src/scalar/func/impls/float64.rs:
```diff
@@ -149,6 +154,7 @@ sqlfunc!(
     #[preserves_uniqueness = false]
     #[inverse = to_unary!(super::CastStringToFloat64)]
+    #[is_monotone = true]
     fn cast_float64_to_string(a: f64) -> String {
```
You're right - I'm not sure what I was thinking with these. I'll roll it back; thanks for the catch!

bkirwi commented on src/expr/src/scalar/func/impls/timestamp.rs:
```diff
@@ -81,6 +84,7 @@ sqlfunc!(
     #[preserves_uniqueness = true]
     #[inverse = to_unary!(super::CastTimestampToTimestampTz)]
+    #[is_monotone = true]
     fn cast_timestamp_tz_to_timestamp(
```
I think this one is correct! Some relevant definitions:

```rust
    /// Returns a view to the naive UTC datetime.
    #[inline]
    pub fn naive_utc(&self) -> NaiveDateTime {
        self.datetime
    }

    /// Returns a view to the naive local datetime.
    #[inline]
    pub fn naive_local(&self) -> NaiveDateTime {
        self.datetime + self.offset.fix()
    }
```

ie. the stored datetime is already in UTC; it's stored alongside a timezone which is applied whenever you ask for the local time. And ordering just compares the UTC datetimes, ignoring the timezone.

ggevay replied:
Oh, ok! I was confused because the `DateTime` struct itself has an `offset` field, so I thought the offset might have different values. (In that case, the offset could flip the comparison, and then it could happen that 
`a.datetime + a.offset < b.datetime + b.offset`, but
`cast_that_forgets_the_offset(a) > cast_that_forgets_the_offset(b)`) 

But reading the [relevant Materialize docs](https://materialize.com/docs/sql/types/timestamp/), it seems the offset is always 0, if I understand correctly? You might want to add a comment there to explain this. Also, I'm not sure whether this behavior might change in the future, i.e., the `offset` might become used, and then someone will need to not forget to adjust `is_monotone`. 

@mjibson, what do you think?
- Is the `offset` field really always 0 here?
- Will it always stay this way? If not, where should we add a comment to avoid the `is_monotone` annotation suddenly becoming invalid in the future? Or if the `offset` field would become used, then this cast function would also start including the `offset` field (e.g., call `naive_local` instead of `naive_utc`), and then the cast would stay monotone?

ggevay replied:
> nor comparison even look the offset field at all

Ah, I get it now, this was the critical part that I missed. Sorry for being slow!

bkirwi replied:
Not at all... this stuff is pretty confusing!

bkirwi commented:
Rebased past a conflict. I've also added a couple cases in the macro and to remove the extra function-level annotations; I'll follow up with the changes to the macro if we can find a nice way to do it.

I'll get this in to avoid more conflicts, but as always happy to follow up if there's additional feedback!
